homeassistant/components/decora_wifi/common.py

```diff
@@ -0,0 +1,156 @@
+"""Common code for decora_wifi."""
+
+from __future__ import annotations
+
+import logging
+
+from decora_wifi import DecoraWiFiSession
+from decora_wifi.models.iot_switch import IotSwitch
+from decora_wifi.models.person import Person
+from decora_wifi.models.residence import Residence
+from decora_wifi.models.residential_account import ResidentialAccount
+
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.dispatcher import async_dispatcher_connect
+from homeassistant.helpers.entity import Entity
+
+from .const import DOMAIN, LIGHT_DOMAIN, PLATFORMS
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class DecoraWifiLoginFailed(Exception):
+    """Raised when DecoraWifiPlatform.login() fails to log in."""
+
+
+class DecoraWifiCommFailed(Exception):
+    """Raised when DecoraWifiPlatform.login() fails to communicate with the myLeviton Service."""
+
+
+class DecoraWifiPlatform:
+    """Class to hold decora_wifi platform sessions and related methods."""
+
+    def __init__(self, email: str, password: str) -> None:
+        """Iniialize session holder."""
+        self._session = DecoraWiFiSession()
+        self._email = email
+        self._password = password
+        self._iot_switches: dict[str, IotSwitch] = {
+            platform: [] for platform in PLATFORMS
+        }
+        self._loggedin = False
+
+        self._apilogin()
+        self._apigetdevices()
+
+    def __del__(self):
+        """Clean up the session on object deletion."""
+        self.apilogout()
+
+    @property
+    def lights(self) -> list[IotSwitch]:
+        """Get the lights."""
+        return self._iot_switches[LIGHT_DOMAIN]
+
+    @property
+    def activeplatforms(self) -> list[str]:
+        """Get the list of platforms which have devices defined."""
+        return [p for p in PLATFORMS if self._iot_switches[p]]
+
+    def _apilogin(self):
+        """Log in to decora_wifi session."""
+        try:
+            success = self._session.login(self._email, self._password)
+
+            # If the call to the decora_wifi API's session.login returns None, there was a problem with the credentials.
+            if success is None:
+                raise DecoraWifiLoginFailed
+            self._loggedin = True
+        except ValueError as exc:
```

Comment:
`ValueError` seems like a bug in the library.  Ideally it will throw a named exception instead

Reply:
Definitely agree that it would be better if the upstream threw a named exception here; [this is just how they chose to write it](https://github.com/tlyakhov/python-decora_wifi/blob/master/decora_wifi/__init__.py#L35).  I'm not a maintainer on it, just a user who owns a couple of the Leviton devices. I'll try submitting a PR to get that improved.

Reply:
https://github.com/tlyakhov/python-decora_wifi/pull/19